File: webserial/protocol.js
// RobStride 私有プロトコル (CAN 2.0B 拡張フレーム) と
// RobStride 純正 USB-CAN アダプタ (AT モード) のシリアルフレーム変換。
//
// backend/protocol.py の移植。実装根拠は docs/protocol.md を参照。
//   シリアルフレーム = 'AT' + BE32((extId << 3) | 0x04) + DLC + data[DLC] + CRLF

import { PARAMS, PLACEHOLDERS, PROFILES, DEFAULT_PROFILE, GROUP_LABELS } from './tables.js';

export const HEADER = Uint8Array.of(0x41, 0x54);   // 'AT'
export const TAIL = Uint8Array.of(0x0d, 0x0a);     // CRLF
export const EXT_FLAG = 0x04;                      // 拡張フレーム (IDE) ビット
export const DEFAULT_HOST_ID = 0xfd;

export const CommType = {
  GET_DEVICE_ID: 0x00,
  MOTION_CONTROL: 0x01,
  FEEDBACK: 0x02,
  ENABLE: 0x03,
  STOP: 0x04,
  SET_ZERO: 0x06,
  SET_CAN_ID: 0x07,
  READ_PARAM: 0x11,
  WRITE_PARAM: 0x12,
  FAULT_FEEDBACK: 0x15,
  SAVE: 0x16,
  SET_BAUD: 0x17,
  ACTIVE_REPORT: 0x18,
  SET_PROTOCOL: 0x19,
  READ_VERSION: 0x1a,
};

export const COMM_NAMES = {
  0x00: 'デバイスID', 0x01: '運動制御', 0x02: 'フィードバック', 0x03: '運転許可',
  0x04: '停止', 0x06: '原点設定', 0x07: 'CAN_ID変更', 0x11: 'パラメータ読出',
  0x12: 'パラメータ書込', 0x15: '故障通知', 0x16: '保存', 0x17: 'ボーレート変更',
  0x18: '能動送信', 0x19: 'プロトコル変更', 0x1a: 'バージョン',
};

// フィードバックフレーム bit21-16 の故障ビット (LSB = bit16)
export const FEEDBACK_FAULT_BITS = [
  [0, '低電圧'], [1, '三相電流異常'], [2, '過温度'],
  [3, '磁気エンコーダ異常'], [4, 'ロック/過負荷'], [5, '未キャリブレーション'],
];

// 通信タイプ 0x15 Byte0-3 の故障ビット
export const FAULT_FRAME_BITS = [
  [0, 'モータ過温度 (既定 145℃)'], [1, 'ドライバチップ故障'], [2, '低電圧故障'],
  [3, '過電圧故障'], [4, 'B 相電流サンプリング過電流'], [5, 'C 相電流サンプリング過電流'],
  [7, 'エンコーダ未キャリブレーション'], [8, 'ハードウェア識別故障'],
  [9, '位置初期化故障'], [14, 'ロック過負荷アルゴリズム保護'],
  [16, 'A 相電流サンプリング過電流'],
];

const MODE_NAMES = { 0: 'Reset', 1: 'Cali', 2: 'Run' };

// --------------------------------------------------------------------------
// 拡張 ID
// --------------------------------------------------------------------------

export function packExtId(commType, data2, targetId) {
  // 29bit を扱うので符号なしにするため >>> 0 を通す
  return ((((commType & 0x1f) << 24) | ((data2 & 0xffff) << 8) | (targetId & 0xff)) >>> 0);
}

export function unpackExtId(extId) {
  return {
    commType: (extId >>> 24) & 0x1f,
    data2: (extId >>> 8) & 0xffff,
    targetId: extId & 0xff,
  };
}

// --------------------------------------------------------------------------
// スケーリング (マニュアルの float_to_uint と同一式: 分母は 2^bits - 1)
// --------------------------------------------------------------------------

export function floatToUint(x, xMin, xMax, bits = 16) {
  const span = xMax - xMin;
  const v = Math.min(Math.max(Number(x), xMin), xMax);
  return Math.trunc(((v - xMin) * ((1 << bits) - 1)) / span);
}

export function uintToFloat(v, xMin, xMax, bits = 16) {
  const span = xMax - xMin;
  return xMin + (v * span) / ((1 << bits) - 1);
}

// --------------------------------------------------------------------------
// フレーム
// --------------------------------------------------------------------------

export function encodeFrame(extId, data, extended = true) {
  if (data.length > 8) throw new Error('CAN のデータ長は最大 8 バイトです');
  const mask = extended ? 0x1fffffff : 0x7ff;
  const wire = (((extId & mask) * 8) + (extended ? EXT_FLAG : 0)) >>> 0;
  const out = new Uint8Array(7 + data.length + 2);
  out[0] = 0x41; out[1] = 0x54;
  out[2] = (wire >>> 24) & 0xff;
  out[3] = (wire >>> 16) & 0xff;
  out[4] = (wire >>> 8) & 0xff;
  out[5] = wire & 0xff;
  out[6] = data.length;
  out.set(data, 7);
  out[7 + data.length] = 0x0d;
  out[8 + data.length] = 0x0a;
  return out;
}

export function buildFrame(commType, data2, targetId, data = new Uint8Array(8)) {
  return encodeFrame(packExtId(commType, data2, targetId), data);
}

export function hex(bytes) {
  return Array.from(bytes, (b) => b.toString(16).padStart(2, '0').toUpperCase()).join(' ');
}

/** フレームとして「構造だけ」合っているか (ヘッダ化けの診断用)。 */
export function looksLikeFrame(raw) {
  return raw.length === 17 && raw[15] === 0x0d && raw[16] === 0x0a && raw[6] === 8;
}

/** バイトストリームからフレームを取り出す耐ノイズパーサ。 */
export class FrameParser {
  static MAX_BUFFER = 4096;

  constructor() { this.buf = []; }

  /** -> [{kind:'frame'|'junk', ...}] */
  feed(chunk) {
    for (const b of chunk) this.buf.push(b);
    const events = [];
    const junk = [];

    for (;;) {
      let start = -1;
      for (let i = 0; i + 1 < this.buf.length; i += 1) {
        if (this.buf[i] === 0x41 && this.buf[i + 1] === 0x54) { start = i; break; }
      }
      if (start < 0) {
        // ヘッダ無し。'A' で終わる可能性があるので 1 バイトだけ残す
        const keep = this.buf.length && this.buf[this.buf.length - 1] === 0x41 ? 1 : 0;
        if (this.buf.length > keep) {
          junk.push(...this.buf.splice(0, this.buf.length - keep));
        }
        break;
      }
      if (start) junk.push(...this.buf.splice(0, start));
      if (this.buf.length < 7) break;

      const dlc = this.buf[6];
      if (dlc > 8) { junk.push(...this.buf.splice(0, 2)); continue; }
      const total = 7 + dlc + 2;
      if (this.buf.length < total) {
        if (this.buf.length > FrameParser.MAX_BUFFER) { junk.push(...this.buf.splice(0, 2)); continue; }
        break;
      }
      if (this.buf[total - 2] !== 0x0d || this.buf[total - 1] !== 0x0a) {
        junk.push(...this.buf.splice(0, 2));   // 偽ヘッダ。2 バイト進めて再同期
        continue;
      }

      const raw = this.buf.splice(0, total);
      const wire = ((raw[2] << 24) | (raw[3] << 16) | (raw[4] << 8) | raw[5]) >>> 0;
      const extended = Boolean(wire & EXT_FLAG);
      const idMask = extended ? 0x1fffffff : 0x7ff;
      events.push({
        kind: 'frame',
        frame: new CanFrame(Math.floor(wire / 8) & idMask, Uint8Array.from(raw.slice(7, 7 + dlc)), extended),
      });
    }

    if (junk.length) events.push({ kind: 'junk', data: Uint8Array.from(junk) });
    return events;
  }
}

export class CanFrame {
  constructor(extId, data, extended = true) {
    this.extId = extId >>> 0;
    this.data = data;
    this.extended = extended;
  }

  get commType() { return (this.extId >>> 24) & 0x1f; }
  get data2() { return (this.extId >>> 8) & 0xffff; }
  get targetId() { return this.extId & 0xff; }
  /** 応答フレームでは bit15-8 が送信元モータの CAN_ID */
  get motorId() { return (this.extId >>> 8) & 0xff; }
  get idHex() { return this.extId.toString(16).padStart(8, '0').toUpperCase(); }
  get dataHex() { return hex(this.data); }
  toString() { return `${this.idHex}${this.extended ? '' : 'S'}#${this.dataHex}`; }
}

// --------------------------------------------------------------------------
// デコード
// --------------------------------------------------------------------------

export function decodeFeedback(frame, pr) {
  if (frame.data.length < 8) throw new Error('フィードバックフレームは 8 バイト必要です');
  const dv = new DataView(frame.data.buffer, frame.data.byteOffset, 8);
  const faultBits = (frame.extId >>> 16) & 0x3f;
  const mode = (frame.extId >>> 22) & 0x03;
  return {
    motorId: frame.motorId,
    angle: uintToFloat(dv.getUint16(0), pr.p_min, pr.p_max),
    velocity: uintToFloat(dv.getUint16(2), pr.v_min, pr.v_max),
    torque: uintToFloat(dv.getUint16(4), pr.t_min, pr.t_max),
    temperature: dv.getUint16(6) / 10,
    mode,
    modeName: MODE_NAMES[mode] ?? `不明(${mode})`,
    faults: FEEDBACK_FAULT_BITS.filter(([b]) => faultBits & (1 << b)).map(([, n]) => n),
    faultBits,
  };
}

export function decodeFaultFrame(frame) {
  const dv = new DataView(frame.data.buffer, frame.data.byteOffset, frame.data.length);
  const fault = frame.data.length >= 4 ? dv.getUint32(0, true) : 0;
  const warn = frame.data.length >= 8 ? dv.getUint32(4, true) : 0;
  return {
    motorId: frame.motorId,
    faultValue: fault,
    warningValue: warn,
    faults: FAULT_FRAME_BITS.filter(([b]) => fault & (1 << b)).map(([, n]) => n),
    warnings: warn & 1 ? ['モータ過温度警告 (既定 135℃)'] : [],
  };
}

/**
 * 通信タイプ 1 の (extId, data) を作る。
 * トルクは 8 バイトのデータ領域ではなく拡張 ID のデータ領域2 に入る。
 * データ領域はすべて上位バイト先行 (ビッグエンディアン)。
 */
export function encodeMotionControl({ torque, position, velocity, kp, kd, motorId, pr }) {
  const extId = packExtId(CommType.MOTION_CONTROL,
    floatToUint(torque, pr.t_min, pr.t_max), motorId);
  const data = new Uint8Array(8);
  const dv = new DataView(data.buffer);
  dv.setUint16(0, floatToUint(position, pr.p_min, pr.p_max));
  dv.setUint16(2, floatToUint(velocity, pr.v_min, pr.v_max));
  dv.setUint16(4, floatToUint(kp, 0, pr.kp_max));
  dv.setUint16(6, floatToUint(kd, 0, pr.kd_max));
  return { extId, data };
}

// --------------------------------------------------------------------------
// パラメータ値 <-> 4 バイト枠 (Byte4-7, リトルエンディアン)
// --------------------------------------------------------------------------

export function encodeValue(type, value) {
  const out = new Uint8Array(4);
  const dv = new DataView(out.buffer);
  if (type === 'float') dv.setFloat32(0, Number(value), true);
  else if (type === 'uint8') out[0] = Number(value) & 0xff;
  else if (type === 'uint16') dv.setUint16(0, Number(value) & 0xffff, true);
  else if (type === 'uint32') dv.setUint32(0, Number(value) >>> 0, true);
  else throw new Error(`未知のパラメータ型: ${type}`);
  return out;
}

export function decodeValue(type, raw) {
  const buf = new Uint8Array(4);
  buf.set(raw.slice(0, 4));
  const dv = new DataView(buf.buffer);
  if (type === 'float') return dv.getFloat32(0, true);
  if (type === 'uint8') return buf[0];
  if (type === 'uint16') return dv.getUint16(0, true);
  if (type === 'uint32') return dv.getUint32(0, true);
  throw new Error(`未知のパラメータ型: ${type}`);
}

// --------------------------------------------------------------------------
// プロファイルとパラメータ表
// --------------------------------------------------------------------------

export function getProfile(key) {
  return PROFILES[String(key ?? '').toUpperCase()] ?? PROFILES[DEFAULT_PROFILE];
}

/** レンジのプレースホルダを機種プロファイルで解決する (params.py の resolve と同じ)。 */
export function resolveRange(value, pr, positive) {
  if (typeof value !== 'string') return value;
  const P_ = PLACEHOLDERS;
  if (value === P_.V_LIM) return positive ? pr.v_max : pr.v_min;
  if (value === P_.V_POS) return pr.v_max;
  if (value === P_.T_POS) return pr.t_max;
  if (value === P_.I_LIM || value === P_.I_POS) {
    if (pr.i_max === null || pr.i_max === undefined) return null;  // 不明: 制限しない
    if (value === P_.I_POS) return pr.i_max;
    return positive ? pr.i_max : -pr.i_max;
  }
  return null;
}

export function schemaFor(pr) {
  return PARAMS.map((p) => ({
    ...p,
    groupLabel: GROUP_LABELS[p.group],
    min: resolveRange(p.min, pr, false),
    max: resolveRange(p.max, pr, true),
  }));
}

export const PARAMS_BY_NAME = Object.fromEntries(PARAMS.map((p) => [p.name, p]));
export const PARAMS_BY_INDEX = Object.fromEntries(PARAMS.map((p) => [p.index, p]));

File: backend/protocol.py
"""RobStride 私有プロトコル (CAN 2.0B 拡張フレーム) と
RobStride 純正 USB-CAN アダプタ (AT モード) のシリアルフレーム変換。

出典 (すべて公式マニュアルで確認済み):
  * 29bit 拡張 ID = [bit28-24 通信タイプ][bit23-8 データ領域2][bit7-0 宛先 CAN_ID]
  * シリアルフレーム = 'AT' + BE32((ext_id << 3) | 0x04) + DLC + data[DLC] + CRLF
    マニュアル記載例:
      41 54 90 07 e8 0c 08 05 70 00 00 01 00 00 00 0d 0a
      -> 0x9007E80C >> 3 == 0x1200FD01
         type=0x12(書込) / host=0x00FD / motor=0x01 / index=0x7005 / value=1
"""

from __future__ import annotations

import struct
from dataclasses import dataclass, field
from enum import IntEnum
from typing import Iterator, List, Tuple

HEADER = b"AT"
TAIL = b"\r\n"

#: 拡張フレームであることを示す下位 3bit のフラグ (IDE ビット)
EXT_FLAG = 0x04

#: ホスト側 CAN_ID の既定値。マニュアルの例が 0xFD を使用している。
DEFAULT_HOST_ID = 0xFD


class CommType(IntEnum):
    """通信タイプ (拡張 ID の bit28-24)。"""

    GET_DEVICE_ID = 0x00      # デバイス ID / MCU 固有 ID 取得
    MOTION_CONTROL = 0x01     # 運動制御モード指令
    FEEDBACK = 0x02           # モータフィードバック
    ENABLE = 0x03             # 運転許可
    STOP = 0x04               # 停止 (Byte0=1 で故障クリア)
    SET_ZERO = 0x06           # 機械原点設定
    SET_CAN_ID = 0x07         # CAN_ID 変更 (即時反映)
    READ_PARAM = 0x11         # 単一パラメータ読出
    WRITE_PARAM = 0x12        # 単一パラメータ書込 (電源断で消失)
    FAULT_FEEDBACK = 0x15     # 故障フィードバック
    SAVE = 0x16               # パラメータ保存 (0x20xx 系を不揮発化)
    SET_BAUD = 0x17           # CAN ボーレート変更 (再投入で反映)
    ACTIVE_REPORT = 0x18      # 能動送信の ON/OFF・その報告フレーム
    SET_PROTOCOL = 0x19       # プロトコル切替 (再投入で反映)
    READ_VERSION = 0x1A       # バージョン読出


class RunMode(IntEnum):
    """0x7005 run_mode の値。"""

    OPERATION = 0   # 運動制御モード (MIT 風 5 パラメータ)
    POSITION_PP = 1 # 位置モード (PP: 台形加減速)
    SPEED = 2       # 速度モード
    CURRENT = 3     # 電流モード
    POSITION_CSP = 5  # 位置モード (CSP: 速度制限付きサイクリック)


class MotorMode(IntEnum):
    """フィードバックフレーム bit23-22 のモードステータス。"""

    RESET = 0
    CALIBRATION = 1
    RUN = 2


#: フィードバックフレーム bit21-16 の故障ビット (LSB = bit16)
FEEDBACK_FAULT_BITS: Tuple[Tuple[int, str], ...] = (
    (0, "低電圧"),
    (1, "三相電流異常"),
    (2, "過温度"),
    (3, "磁気エンコーダ異常"),
    (4, "ロック/過負荷"),
    (5, "未キャリブレーション"),
)

#: 通信タイプ 0x15 (故障フィードバック) Byte0-3 の故障ビット
FAULT_FRAME_BITS: Tuple[Tuple[int, str], ...] = (
    (0, "モータ過温度 (既定 145℃)"),
    (1, "ドライバチップ故障"),
    (2, "低電圧故障"),
    (3, "過電圧故障"),
    (4, "B 相電流サンプリング過電流"),
    (5, "C 相電流サンプリング過電流"),
    (7, "エンコーダ未キャリブレーション"),
    (8, "ハードウェア識別故障"),
    (9, "位置初期化故障"),
    (14, "ロック過負荷アルゴリズム保護"),
    (16, "A 相電流サンプリング過電流"),
)

#: 通信タイプ 0x15 Byte4-7 の警告ビット
WARNING_FRAME_BITS: Tuple[Tuple[int, str], ...] = (
    (0, "モータ過温度警告 (既定 135℃)"),
)


# --------------------------------------------------------------------------
# 拡張 ID の組立/分解
# --------------------------------------------------------------------------

def pack_ext_id(comm_type: int, data2: int, target_id: int) -> int:
    return ((int(comm_type) & 0x1F) << 24) | ((int(data2) & 0xFFFF) << 8) | (int(target_id) & 0xFF)


def unpack_ext_id(ext_id: int) -> Tuple[int, int, int]:
    """-> (通信タイプ, データ領域2, 宛先 CAN_ID)"""
    return (ext_id >> 24) & 0x1F, (ext_id >> 8) & 0xFFFF, ext_id & 0xFF


# --------------------------------------------------------------------------
# 値のスケーリング (マニュアルの float_to_uint と同一式: 分母は 2^bits - 1)
# --------------------------------------------------------------------------

def float_to_uint(x: float, x_min: float, x_max: float, bits: int = 16) -> int:
    span = x_max - x_min
    x = min(max(float(x), x_min), x_max)
    return int((x - x_min) * ((1 << bits) - 1) / span)


def uint_to_float(v: int, x_min: float, x_max: float, bits: int = 16) -> float:
    span = x_max - x_min
    return x_min + (v * span) / ((1 << bits) - 1)


# --------------------------------------------------------------------------
# CAN フレーム
# --------------------------------------------------------------------------

@dataclass(frozen=True)
class CanFrame:
    ext_id: int
    data: bytes
    extended: bool = True

    @property
    def comm_type(self) -> int:
        return (self.ext_id >> 24) & 0x1F

    @property
    def data2(self) -> int:
        return (self.ext_id >> 8) & 0xFFFF

    @property
    def target_id(self) -> int:
        return self.ext_id & 0xFF

    @property
    def motor_id(self) -> int:
        """応答フレームでは bit15-8 が送信元モータの CAN_ID。"""
        return (self.ext_id >> 8) & 0xFF

    def hex(self) -> str:
        return " ".join(f"{b:02X}" for b in self.data)

    def __str__(self) -> str:  # ログ表示用
        kind = "" if self.extended else "S"
        return f"{self.ext_id:08X}{kind}#{self.hex()}"


def encode_frame(ext_id: int, data: bytes, extended: bool = True) -> bytes:
    """CAN フレームを USB-CAN アダプタのシリアルフレームへ。

    RobStride 私有プロトコルは拡張フレームのみを使うが、MIT プロトコルへ
    切り替えたモータは 11bit 標準フレームで通信する。標準フレームでは
    IDE ビット (0x04) を落とす — マニュアルに明示的な記載はないため、
    拡張フレームの記載から類推した実装。
    """
    if len(data) > 8:
        raise ValueError("CAN のデータ長は最大 8 バイトです")
    mask = 0x1FFFFFFF if extended else 0x7FF
    wire = ((ext_id & mask) << 3) | (EXT_FLAG if extended else 0x00)
    return HEADER + struct.pack(">I", wire) + bytes([len(data)]) + bytes(data) + TAIL


def build_frame(comm_type: int, data2: int, target_id: int, data: bytes = b"\x00" * 8) -> bytes:
    return encode_frame(pack_ext_id(comm_type, data2, target_id), data)


class FrameParser:
    """バイトストリームからシリアルフレームを取り出す耐ノイズパーサ。

    アダプタは AT コマンドへの ASCII 応答 (``OK`` など) も返すため、
    フレームとして解釈できなかったバイト列は ``junk`` イベントとして
    生バイトのまま通知する。
    """

    #: 未同期バッファがこの長さを超えたら先頭を捨てる
    MAX_BUFFER = 4096

    def __init__(self) -> None:
        self._buf = bytearray()

    def feed(self, chunk: bytes) -> List[Tuple[str, object]]:
        """-> [('frame', CanFrame) | ('junk', bytes), ...]

        フレームとして解釈できなかったバイト列は ``junk`` として **生バイトのまま**
        返す。ASCII に変換してしまうと 0x80 以上のバイトが失われ、ボーレート違いや
        別フレーム形式の診断ができなくなるため。
        """
        self._buf.extend(chunk)
        events: List[Tuple[str, object]] = []
        junk = bytearray()

        while True:
            start = self._buf.find(HEADER)
            if start < 0:
                # ヘッダが無い。'A' で終わる可能性があるので 1 バイトだけ残す。
                keep = 1 if self._buf.endswith(b"A") else 0
                if len(self._buf) > keep:
                    junk.extend(self._buf[: len(self._buf) - keep])
                    del self._buf[: len(self._buf) - keep]
                break

            if start:
                junk.extend(self._buf[:start])
                del self._buf[:start]

            if len(self._buf) < 7:  # AT + id(4) + dlc(1) が揃うまで待つ
                break
            dlc = self._buf[6]
            if dlc > 8:
                junk.extend(self._buf[:2])
                del self._buf[:2]
                continue
            total = 7 + dlc + len(TAIL)
            if len(self._buf) < total:
                if len(self._buf) > self.MAX_BUFFER:
                    junk.extend(self._buf[:2])
                    del self._buf[:2]
                    continue
                break
            if bytes(self._buf[total - 2 : total]) != TAIL:
                # 末尾が合わない = 偽ヘッダ。2 バイト進めて再同期。
                junk.extend(self._buf[:2])
                del self._buf[:2]
                continue

            wire = struct.unpack(">I", bytes(self._buf[2:6]))[0]
            data = bytes(self._buf[7 : 7 + dlc])
            del self._buf[:total]
            extended = bool(wire & EXT_FLAG)
            raw_id = (wire >> 3) & (0x1FFFFFFF if extended else 0x7FF)
            events.append(("frame", CanFrame(ext_id=raw_id, data=data, extended=extended)))

        if junk:
            events.append(("junk", bytes(junk)))
        return events


# --------------------------------------------------------------------------
# フレームのデコード
# --------------------------------------------------------------------------

@dataclass
class Feedback:
    """通信タイプ 0x02 / 0x18 のフィードバック。"""

    motor_id: int
    angle: float          # rad
    velocity: float       # rad/s
    torque: float         # Nm
    temperature: float    # ℃
    mode: int
    mode_name: str
    faults: List[str] = field(default_factory=list)
    fault_bits: int = 0

    def as_dict(self) -> dict:
        return {
            "motor_id": self.motor_id,
            "angle": self.angle,
            "velocity": self.velocity,
            "torque": self.torque,
            "temperature": self.temperature,
            "mode": self.mode,
            "mode_name": self.mode_name,
            "faults": self.faults,
            "fault_bits": self.fault_bits,
        }


_MODE_NAMES = {0: "Reset", 1: "Cali", 2: "Run"}


def decode_feedback(frame: CanFrame, p_min: float, p_max: float,
                    v_min: float, v_max: float,
                    t_min: float, t_max: float) -> Feedback:
    """フィードバックフレームを物理量へ変換する。

    スケーリングレンジはモータ機種ごとに異なるため呼び出し側が渡す。
    """
    if len(frame.data) < 8:
        raise ValueError("フィードバックフレームは 8 バイト必要です")
    angle_raw, vel_raw, tor_raw, temp_raw = struct.unpack(">HHHH", frame.data[:8])
    fault_bits = (frame.ext_id >> 16) & 0x3F
    mode = (frame.ext_id >> 22) & 0x03
    return Feedback(
        motor_id=frame.motor_id,
        angle=uint_to_float(angle_raw, p_min, p_max),
        velocity=uint_to_float(vel_raw, v_min, v_max),
        torque=uint_to_float(tor_raw, t_min, t_max),
        temperature=temp_raw / 10.0,
        mode=mode,
        mode_name=_MODE_NAMES.get(mode, f"不明({mode})"),
        faults=[name for bit, name in FEEDBACK_FAULT_BITS if fault_bits & (1 << bit)],
        fault_bits=fault_bits,
    )


def decode_fault_frame(frame: CanFrame) -> dict:
    """通信タイプ 0x15 のデコード。"""
    fault = warn = 0
    if len(frame.data) >= 4:
        fault = struct.unpack("<I", frame.data[0:4])[0]
    if len(frame.data) >= 8:
        warn = struct.unpack("<I", frame.data[4:8])[0]
    return {
        "motor_id": frame.motor_id,
        "fault_value": fault,
        "warning_value": warn,
        "faults": [n for b, n in FAULT_FRAME_BITS if fault & (1 << b)],
        "warnings": [n for b, n in WARNING_FRAME_BITS if warn & (1 << b)],
    }


def encode_motion_control(torque: float, position: float, velocity: float,
                          kp: float, kd: float, motor_id: int,
                          p_min: float, p_max: float, v_min: float, v_max: float,
                          t_min: float, t_max: float,
                          kp_max: float, kd_max: float) -> Tuple[int, bytes]:
    """通信タイプ 0x01 の (ext_id, data) を作る。

    トルクは 8 バイトのデータ領域ではなく拡張 ID のデータ領域2 に入る。
    データ領域はすべて上位バイト先行 (ビッグエンディアン)。
    """
    ext_id = pack_ext_id(
        CommType.MOTION_CONTROL,
        float_to_uint(torque, t_min, t_max),
        motor_id,
    )
    data = struct.pack(
        ">HHHH",
        float_to_uint(position, p_min, p_max),
        float_to_uint(velocity, v_min, v_max),
        float_to_uint(kp, 0.0, kp_max),
        float_to_uint(kd, 0.0, kd_max),
    )
    return ext_id, data


def iter_hex(data: bytes) -> Iterator[str]:
    for b in data:
        yield f"{b:02X}"


def looks_like_frame(raw: bytes) -> bool:
    """AT フレームとして「構造だけ」合っているかを判定する。

    ヘッダが化けていてもフレーム長・DLC・末尾 CRLF が揃っていれば True。
    リンク品質の診断に使う (デコードはしない — ID 部も化けている可能性が
    あるため、値として信用してはいけない)。
    """
    return (len(raw) == 7 + 8 + len(TAIL)
            and raw[-2:] == TAIL
            and raw[6] == 8)
